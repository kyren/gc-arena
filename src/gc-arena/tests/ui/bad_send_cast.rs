use std::rc::Rc;

use gc_arena::{unsize, Collect, Gc, MutationContext, Rootable};

trait TestTrait<'gc> {}

#[derive(Collect)]
#[collect(no_drop)]
struct TestStruct<'gc>(Gc<'gc, Rc<i32>>);

impl<'gc> TestTrait<'gc> for TestStruct<'gc> {}

fn test<'gc>(_mc: MutationContext<'gc, '_>, test: Gc<'gc, TestStruct<'gc>>) {
    let _p: Gc<'gc, dyn TestTrait<'gc> + Send> = unsize!(
        test,
        Rootable!['a => Gc<'a, TestStruct<'a>>] =>
        Rootable!['a => Gc<'a, dyn TestTrait<'a> + Send>]
    );
}

fn main() {}
