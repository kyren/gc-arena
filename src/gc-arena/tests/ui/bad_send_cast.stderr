error[E0277]: `Rc<i32>` cannot be sent between threads safely
  --> tests/ui/bad_send_cast.rs:14:50
   |
14 |       let _p: Gc<'gc, dyn TestTrait<'gc> + Send> = unsize!(
   |  __________________________________________________^
15 | |         test,
16 | |         Rootable!['a => Gc<'a, TestStruct<'a>>] =>
17 | |         Rootable!['a => Gc<'a, dyn TestTrait<'a> + Send>]
18 | |     );
   | |_____^ `Rc<i32>` cannot be sent between threads safely
   |
   = help: the trait `Send` is not implemented for `Rc<i32>`
   = note: required for `Gc<'_, Rc<i32>>` to implement `Send`
note: required because it appears within the type `TestStruct<'_>`
  --> tests/ui/bad_send_cast.rs:9:8
   |
9  | struct TestStruct<'gc>(Gc<'gc, Rc<i32>>);
   |        ^^^^^^^^^^
   = note: required for the cast from `TestStruct<'_>` to the object type `dyn TestTrait<'_> + Send`
   = note: this error originates in the macro `unsize` (in Nightly builds, run with -Z macro-backtrace for more info)
